classes/services/PKPFileService.inc.php

```diff
@@ -85,9 +86,10 @@ public function add($from, $to) {
 		if (is_resource($stream)) {
 			fclose($stream);
 		}
+		$mimeDetector = new FinfoMimeTypeDetector();
 		return Capsule::table('files')->insertGetId([
 			'path' => $to,
-			'mimetype' => $this->fs->getMimetype($to),
+			'mimetype' => $mimeDetector->detectMimeType($to, file_get_contents($from)),
```

Comment:
Detect mime type by file's content, fall back to the extension if unsuccessful, e.g., the file doesn't have any internal signature that can be associated with a specific file type.

Reply:
I don't think this is going to work with remote file systems. One of the things that Flysystem provides is an abstraction from the filesystem itself. The file may be hosted on a completely different server, like AWS, and the Flysystem methods will still work. Calling something like `file_get_contents` directly on a path won't work with remote calls without specific PHP configuration, and I don't think that it could ever work when the remote file system requires authentication.

Reply:
Yes, in these cases it falls back to the mime identification by extension. Or you are thinking about something more sophisticated?

Reply:
I think that's ok, but we need to use Flysystem for the initial check to be compatible with alternative file systems. So I think what we'd need is something like the following:

```php
$mimetype = $this->fs->getMimetype($to);
if ($mimetype === 'text/plain') {
  // Check file extension based on $to for specific cases like CSS/JS
}
```

I'm not the best at this file system stuff, though. @asmecher is there a better way to detect mimetype based on a filename that will be compatible regardless of where the file is stored?

Reply:
After exploring the library and associated code, I think I understand more about this. But just to double-check, the variable `$from` points to the file in the temporary file folder; temporary files can be stored remotely, thus retrieving the file content with `file_get_content` may not work in this case. My first assumption was that Flysystem abstractions are used here only to store files remotely, so I'm a bit confused.

Reply:
Or you mean just general implication of the `PKPFileService::add` method which could accept remote path as a first parameter, not just path to the temporary file on the local system, and the file abstraction should handle it?

Reply:
All of Flysystem's interactions are with remote files, so not just writing and reading files, but also the mimetype check and other methods in its API.

In this specific case, I think you're right that `file_get_content` will work when using `$from`. And this method probably needs some work to be able to support adding files from an alternate source. But ideally we'd like to separate this out so that there's not such a dependency even in the original file that this is being created from.

Alec will understand the limitations here better...

Reply:
> Or you mean just general implication of the PKPFileService::add method which could accept remote path as a first parameter, not just path to the temporary file on the local system, and the file abstraction should handle it?

Yeah. Looking at the method, i suspect the `fopen` thing and the streaming it that way won't allow this. But I think the idea is to support it (imagine very large data files that might be stored remotely, in something like Dataverse, but which we want to track and treat like "files" in our system).

Reply:
@asmecher I think this is something that needs to make it into 3.3. Should Vitality go ahead with this approach for CSS and JS files?

```php
$mimetype = $this->fs->getMimetype($to);
if ($mimetype === 'text/plain') {
  // Update mimetype if extension is css or js
}
```

Reply:
Yes, @Vitaliy-1, please do. Our current list of exceptions (in 3.2.x) is implemented in [`PKPString::mime_content_type`](https://github.com/pkp/pkp-lib/blob/stable-3_2_1/classes/core/PKPString.inc.php#L351..L362); the most important frequently mismatched types are CSS and HTML.